Framework/MDAlgorithms/src/Integrate3DEvents.cpp

```diff
@@ -69,6 +72,266 @@ void Integrate3DEvents::addEvents(
+    const std::tuple<double, double, double> &libPeak, const V3D &center,
+    double &inti, double &sigi) {
+
+  inti = 0.0; // default values, in case something
+  sigi = 0.0; // is wrong with the peak.
+
+  auto result = getEvents(center);
+  if (!result)
+    return boost::make_shared<NoShape>();
+
+  const auto &events = result.get();
+
+  const auto &directions = shape->directions();
+  const auto &abcBackgroundInnerRadii = shape->abcRadiiBackgroundInner();
+  const auto &abcBackgroundOuterRadii = shape->abcRadiiBackgroundOuter();
+  const auto &abcRadii = shape->abcRadii();
+
+  const auto max_sigma = std::get<2>(libPeak);
+  auto rValues = calculateRadiusFactors(params, max_sigma);
+
+  const auto isPeakOnDetector =
+      correctForDetectorEdges(rValues, params.E1Vectors, center, abcRadii,
+                              abcBackgroundInnerRadii, abcBackgroundOuterRadii);
+
+  if (!isPeakOnDetector)
+    return shape;
+
+  const double r1 = std::get<0>(rValues), r2 = std::get<1>(rValues),
+               r3 = std::get<2>(rValues);
+
+  // integrate
+  double backgrd = numInEllipsoidBkg(
+      events, directions, abcBackgroundOuterRadii, abcBackgroundInnerRadii);
+  double peak_w_back = numInEllipsoid(events, directions, abcRadii);
+  double ratio = pow(r1, 3) / (pow(r3, 3) - pow(r2, 3));
+
+  const auto frac = std::get<0>(libPeak);
+  const auto fracError = std::get<1>(libPeak);
+
+  inti = peak_w_back - ratio * backgrd;
+  sigi = inti + ratio * ratio * backgrd;
+
+  // correct for fractional intensity
+  sigi = sigi / pow(inti, 2);
+  sigi += pow((fracError / frac), 2);
+
+  inti = inti * frac;
+  sigi = sqrt(sigi) * inti;
+
+  if (inti < 0) {
+    inti = 0;
+    sigi = 0;
+  }
+
+  return shape;
+}
+
+double Integrate3DEvents::estimateSignalToNoiseRatio(
+    const IntegrationParameters &params, const V3D &center) {
+
+  auto result = getEvents(center);
+  if (!result)
+    return .0;
+
+  const auto &events = result.get();
+  if (events.empty())
+    return .0;
+
+  DblMatrix cov_matrix(3, 3);
+  makeCovarianceMatrix(events, cov_matrix, params.regionRadius);
+
+  std::vector<V3D> eigen_vectors;
+  getEigenVectors(cov_matrix, eigen_vectors);
+
+  std::vector<double> sigmas;
+  for (int i = 0; i < 3; i++) {
+    sigmas.push_back(stdDev(events, eigen_vectors[i], params.regionRadius));
+  }
+
+  const auto max_sigma = *std::max_element(sigmas.begin(), sigmas.end());
+  if (max_sigma == 0)
+    return .0;
+
+  auto rValues = calculateRadiusFactors(params, max_sigma);
+  auto &r1 = std::get<0>(rValues), r2 = std::get<1>(rValues),
+       r3 = std::get<2>(rValues);
+
+  std::vector<double> abcBackgroundOuterRadii, abcBackgroundInnerRadii;
+  std::vector<double> peakRadii;
+  for (int i = 0; i < 3; i++) {
+    abcBackgroundOuterRadii.push_back(r3 * sigmas[i]);
+    abcBackgroundInnerRadii.push_back(r2 * sigmas[i]);
+    peakRadii.push_back(r1 * sigmas[i]);
+  }
+
+  // Background / Peak / Background
```

Comment:
In the comment? There is a way of estimating intensity or signal to noise that is referred to in the literature as background-peak-background.